packages/core/admin/_internal/node/webpack/watch.ts

```diff
@@ -0,0 +1,99 @@
+import os from 'node:os';
+import path from 'node:path';
+import webpackDevMiddleware from 'webpack-dev-middleware';
+import webpackHotMiddleware from 'webpack-hot-middleware';
+import { webpack } from 'webpack';
+import type { BuildContext } from '../createBuildContext';
+import { mergeConfigWithUserConfig, resolveDevelopmentConfig } from './config';
+import { Common } from '@strapi/types';
+
+const watch = async (ctx: BuildContext) => {
+  const config = await resolveDevelopmentConfig(ctx);
+  const finalConfig = await mergeConfigWithUserConfig(config, ctx);
+
+  ctx.logger.debug('Final webpack config:', os.EOL, finalConfig);
+
+  return new Promise((res) => {
+    const compiler = webpack(finalConfig);
+
+    const devMiddleware = webpackDevMiddleware(compiler);
+
+    // @ts-ignore
+    const hotMiddleware = webpackHotMiddleware(compiler, {
+      log: false,
+      path: '/__webpack_hmr',
+    });
+
+    ctx.strapi.server.app.use((ctx, next) => {
+      return new Promise((resolve, reject) => {
+        hotMiddleware(ctx.req, ctx.res, (err) => {
+          if (err) reject(err);
+          else resolve(next());
+        });
+      });
+    });
+
+    ctx.strapi.server.app.use((context, next) => {
+      // wait for webpack-dev-middleware to signal that the build is ready
+      const ready = new Promise((resolve, reject) => {
+        devMiddleware.waitUntilValid(() => {
+          resolve(true);
+        });
+      });
+      // tell webpack-dev-middleware to handle the request
+      const init = new Promise((resolve) => {
+        devMiddleware(
+          context.req,
+          {
+            // @ts-expect-error
+            end: (content) => {
+              // eslint-disable-next-line no-param-reassign
+              context.body = content;
+              resolve(true);
+            },
+            getHeader: context.get.bind(context),
+            // @ts-expect-error
+            setHeader: context.set.bind(context),
+            locals: context.state,
+          },
+          () => resolve(next())
+        );
```

Comment:
Do you know how webpack handles this? I lack the understanding of what this middleware is doing, rather than redirecting admin routes to the dev environment.

Reply:
It looks messier because of the webpack middlewares are express styled as opposed to koa. But basically it serves files emitted from webpack's compiler 🤓